Switch to an improved example conversion method

t0yv0 opened the pull request:
# Stats

Baseline      67 API locations with examples

Provider:     azuread
Success rate: 98.10% (618/630)

Converted 98.10% of csharp examples (103/105)
Converted 98.10% of go examples (103/105)
Converted 99.05% of java examples (104/105)
Converted 98.10% of python examples (103/105)
Converted 98.10% of typescript examples (103/105)
Converted 97.14% of yaml examples (102/105)

Experimental  67 API locations with examples

Provider:     azuread
Success rate: 98.30% (635/646)

Converted 98.15% of csharp examples (106/108)
Converted 98.15% of go examples (106/108)
Converted 98.15% of java examples (106/108)
Converted 98.15% of python examples (106/108)
Converted 98.15% of typescript examples (106/108)
Converted 99.06% of yaml examples (105/106)

# Newly failing examples

Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
Example started failing: #/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2
java 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
python 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
typescript 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
yaml 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
csharp 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
go 2 Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.

# Error 1

6 examples failed with the following error:

```
Duplicate resource "azuread_app_role_assignment" configuration: A azuread_app_role_assignment resource named "example" was already declared at /e69.tf:27,1-49. Resource names must be unique per type in each module.
```

Failures include converting the "#/resources/azuread:index/appRoleAssignment:AppRoleAssignment#2" example with the following HCL to python, typescript, yaml, csharp, go, java:

```
data "azuread_domains" "example" {
  only_initial = true
}

resource "azuread_application" "internal" {
  display_name = "internal"

  app_role {
    allowed_member_types = ["Application", "User"]
    description          = "Admins can perform all task actions"
    display_name         = "Admin"
    enabled              = true
    id                   = "00000000-0000-0000-0000-222222222222"
    value                = "Admin.All"
  }
}

resource "azuread_service_principal" "internal" {
  application_id = azuread_application.internal.application_id
}

resource "azuread_group" "example" {
  display_name     = "example"
  security_enabled = true
}

resource "azuread_app_role_assignment" "example" {
  app_role_id         = azuread_service_principal.internal.app_role_ids["Admin.All"]
  principal_object_id = azuread_group.example.object_id
  resource_object_id  = azuread_service_principal.internal.object_id
}

resource "azuread_user" "example" {
  display_name        = "D. Duck"
  password            = "SecretP@sswd99!"
  user_principal_name = "d.duck@${data.azuread_domains.example.domains.0.domain_name}"
}

resource "azuread_app_role_assignment" "example" {
  app_role_id         = azuread_service_principal.internal.app_role_ids["Admin.All"]
  principal_object_id = azuread_user.example.object_id
  resource_object_id  = azuread_service_principal.internal.object_id
}
```


# Error 2

1 examples failed with the following error:

```
unsupported attribute 'owners'; unknown property 'owners' among [otherMails surname ageGroup consentProvidedForMinor division forcePasswordChange managerId password preferredLanguage department mail officeLocation companyName costCenter disablePasswordExpiration jobTitle showInAddressList streetAddress usageLocation userPrincipalName city givenName postalCode accountEnabled displayName employeeType disableStrongPassword employeeId faxNumber mailNickname state businessPhones country mobilePhone onpremisesImmutableId]
```

Failures include converting the "#/resources/azuread:index/group:Group#2" example with the following HCL to csharp:

```
data "azuread_client_config" "current" {}

resource "azuread_user" "example" {
  display_name        = "J Doe"
  owners              = [data.azuread_client_config.current.object_id]
  password            = "notSecure123"
  user_principal_name = "jdoe@example.com"
}

resource "azuread_group" "example" {
  display_name     = "MyGroup"
  owners           = [data.azuread_client_config.current.object_id]
  security_enabled = true

  members = [
    azuread_user.example.object_id,
    /* more users */
  ]
}
```


# Error 3

1 examples failed with the following error:

```
unsupported attribute 'owners'; unknown property 'owners' among [mailNickname preferredLanguage showInAddressList forcePasswordChange department otherMails surname companyName disablePasswordExpiration employeeType streetAddress city postalCode managerId consentProvidedForMinor mobilePhone ageGroup officeLocation state businessPhones costCenter country division employeeId onpremisesImmutableId password usageLocation accountEnabled userPrincipalName displayName faxNumber givenName jobTitle mail disableStrongPassword]
```

Failures include converting the "#/resources/azuread:index/group:Group#2" example with the following HCL to python:

```
data "azuread_client_config" "current" {}

resource "azuread_user" "example" {
  display_name        = "J Doe"
  owners              = [data.azuread_client_config.current.object_id]
  password            = "notSecure123"
  user_principal_name = "jdoe@example.com"
}

resource "azuread_group" "example" {
  display_name     = "MyGroup"
  owners           = [data.azuread_client_config.current.object_id]
  security_enabled = true

  members = [
    azuread_user.example.object_id,
    /* more users */
  ]
}
```


# Error 4

1 examples failed with the following error:

```
unsupported attribute 'owners'; unknown property 'owners' among [disablePasswordExpiration displayName faxNumber mailNickname managerId preferredLanguage costCenter division businessPhones department employeeType forcePasswordChange mobilePhone officeLocation onpremisesImmutableId ageGroup streetAddress employeeId jobTitle mail city country otherMails password accountEnabled showInAddressList state usageLocation companyName userPrincipalName postalCode disableStrongPassword givenName surname consentProvidedForMinor]
```

Failures include converting the "#/resources/azuread:index/group:Group#2" example with the following HCL to typescript:

```
data "azuread_client_config" "current" {}

resource "azuread_user" "example" {
  display_name        = "J Doe"
  owners              = [data.azuread_client_config.current.object_id]
  password            = "notSecure123"
  user_principal_name = "jdoe@example.com"
}

resource "azuread_group" "example" {
  display_name     = "MyGroup"
  owners           = [data.azuread_client_config.current.object_id]
  security_enabled = true

  members = [
    azuread_user.example.object_id,
    /* more users */
  ]
}
```


# Error 5

1 examples failed with the following error:

```
unsupported attribute 'owners'; unknown property 'owners' among [country mobilePhone password accountEnabled city employeeId surname costCenter managerId preferredLanguage showInAddressList streetAddress usageLocation otherMails businessPhones companyName department disablePasswordExpiration displayName employeeType givenName faxNumber postalCode division officeLocation onpremisesImmutableId consentProvidedForMinor forcePasswordChange jobTitle mailNickname ageGroup disableStrongPassword mail state userPrincipalName]
```

Failures include converting the "#/resources/azuread:index/group:Group#2" example with the following HCL to go:

```
data "azuread_client_config" "current" {}

resource "azuread_user" "example" {
  display_name        = "J Doe"
  owners              = [data.azuread_client_config.current.object_id]
  password            = "notSecure123"
  user_principal_name = "jdoe@example.com"
}

resource "azuread_group" "example" {
  display_name     = "MyGroup"
  owners           = [data.azuread_client_config.current.object_id]
  security_enabled = true

  members = [
    azuread_user.example.object_id,
    /* more users */
  ]
}
```


# Error 6

1 examples failed with the following error:

```
not yet implemented: call notImplemented; Function not yet implemented: Function format not yet implemented
```

Failures include converting the "#/resources/azuread:index/directoryRoleAssignment:DirectoryRoleAssignment#2" example with the following HCL to java:

```
resource "azuread_directory_role" "example" {
  display_name = "Cloud application administrator"
}

resource "azuread_application" "example" {
  display_name = "My Application"
}

data "azuread_user" "example" {
  user_principal_name = "jdoe@example.com"
}

resource "azuread_directory_role_assignment" "example" {
  role_id             = azuread_directory_role.example.template_id
  principal_object_id = data.azuread_user.example.object_id
  directory_scope_id  = format("/%s", azuread_application.example.object_id)
}
```

t0yv0 commented on provider/go.mod:
```diff
@@ -5,8 +5,8 @@ go 1.19
 require (
 	github.com/hashicorp/go-azure-sdk v0.20230511.1094507
 	github.com/hashicorp/terraform-provider-azuread/shim v0.0.0
-	github.com/pulumi/pulumi-terraform-bridge/v3 v3.60.1
```
Changes:

https://github.com/pulumi/pulumi-terraform-bridge/compare/v3.60.1...e3aabd38b8f364a8df8522f334ad787b06e33a42

Should probably upgrade bridge to latest released version first.

t0yv0 commented on Makefile:
```diff
@@ -126,4 +128,12 @@ upstream.rebase:
 	$(shell pulumictl download-binary -n pulumi-language-java -v $(JAVA_GEN_VERSION) -r pulumi/pulumi-java)
 
-.PHONY: development build build_sdks install_go_sdk install_java_sdk install_python_sdk install_sdks only_build build_dotnet build_go build_java build_nodejs build_python clean cleanup help install_dotnet_sdk install_nodejs_sdk install_plugins lint_provider provider test tfgen upstream upstream.finalize upstream.rebase
+.PHONY: development build build_sdks install_go_sdk install_java_sdk install_python_sdk install_sdks only_build build_dotnet build_go build_java build_nodejs build_python clean cleanup help install_dotnet_sdk install_nodejs_sdk install_plugins lint_provider provider test tfgen upstream upstream.finalize upstream.rebase upgrade_pulumi
```
TODO Makefile changes need to go back into ci-mgmt.

t0yv0 commented on provider/go.mod:
```diff
@@ -5,7 +5,7 @@ go 1.21
 	github.com/hashicorp/go-azure-sdk v0.20230511.1094507
 	github.com/hashicorp/terraform-provider-azuread/shim v0.0.0
-	github.com/pulumi/pulumi-terraform-bridge/v3 v3.61.0
+	github.com/pulumi/pulumi-terraform-bridge/v3 v3.61.1-0.20231003220302-e3aabd38b8f3
```
Changes:

https://github.com/pulumi/pulumi-terraform-bridge/compare/v3.61.0...e3aabd38b8f364a8df8522f334ad787b06e33a42

t0yv0 commented:
At one point this seemed like it was flip-flopping on `make tfgen` but I'm having a hard time reproducing. Would need the new converter to be 100% deterministic otherwise we need to relax the code equality checks in CI which would be difficult.

```
  import * as pulumi from "@pulumi/pulumi";
  import * as azuread from "@pulumi/azuread";

  const example = new azuread.User("example", {
+     userPrincipalName: "jdoe@hashicorp.com",
      displayName: "J. Doe",
      mailNickname: "jdoe",
      password: "SecretP@sswd99!",
-     userPrincipalName: "jdoe@hashicorp.com",
  });
```

Frassle commented:
> Duplicate resource - simple duplication error in input data; may accept an option to drop duplicate resources instead of erroring..
> unsupported attribute 'owners'; - that's valid, User does not have 'owners', bad input example

Are you passing "--strict" to "convert"? I'd think as long as strictness was off we'd treat these as warnings and do a best effort translation. Can you raise an issue if that's not working.

> Function format not yet implemented - perhaps a ticket? Where does this go?

https://github.com/pulumi/pulumi-std/issues/17

t0yv0 commented:
Awesome, no I don't think we use "--strict".

t0yv0 commented:
Holding till Oct 13 per request.

t0yv0 commented:
Converter actually deterministic it seems like, and error was on my part in the comparison script, this is now fixed.

Frassle commented:
> Converter actually deterministic it seems like, and error was on my part in the comparison script, this is now fixed.

Just an FYI the converter is designed to be deterministic so if you ever find something in that isn't then you can raise that as a converter bug.

t0yv0 commented:
Awesome. No this one was on me, hastily written scripts made it look non-deterministic. Fixed now.

t0yv0 commented:
The updated version of this is non-deterministic. Filing https://github.com/pulumi/pulumi-converter-terraform/issues/87

t0yv0 commented:
This is now stable 💥 thanks Fraser for the latest fix.

I don't understand though what's going on with the stats. Only java examples seem to be updated by "make build_sdks". The updates don't look like an improvement. The stats though seem to say there's more examples working and getting rendered into the schema, but they don't get rendered into the SDKs? I wonder why is that. I'll trace one of them to find out.

```
go run ~/code/pulumi-terraform-bridge/unstable/scripts/exconverter/main.go --dir "$PWD/cache"                                                                                                                ~/code/pulumi-azuread
Baseline      67 API locations with examples

Provider:     azuread
Success rate: 98.10% (618/630)

Converted 98.10% of csharp examples (103/105)
Converted 98.10% of go examples (103/105)
Converted 99.05% of java examples (104/105)
Converted 98.10% of python examples (103/105)
Converted 98.10% of typescript examples (103/105)
Converted 97.14% of yaml examples (102/105)

Experimental  67 API locations with examples

Provider:     azuread
Success rate: 98.30% (635/646)

Converted 98.15% of csharp examples (106/108)
Converted 98.15% of go examples (106/108)
Converted 98.15% of java examples (106/108)
Converted 98.15% of python examples (106/108)
Converted 98.15% of typescript examples (106/108)
Converted 99.06% of yaml examples (105/106)
```

t0yv0 commented:
OK there is no net new examples here, this is a stats bug. Filed that in https://github.com/pulumi/pulumi-terraform-bridge/issues/1572

t0yv0 commented:
Filed https://github.com/pulumi/pulumi-converter-terraform/issues/88

t0yv0 commented:
Filed https://github.com/pulumi/pulumi-converter-terraform/issues/89 also - the YAML version of 88 seems much more severe, as deploying this would break formerly working YAML examples, to the tune of 200 examples. This unfortunately is a blocker.

t0yv0 commented:
Good to go I think.

VenelinMartinov commented on sdk/dotnet/AccessPackageAssignmentPolicy.cs:
```diff
@@ -30,26 +30,26 @@ namespace Pulumi.AzureAD
     /// return await Deployment.RunAsync(() =&gt; 
     /// {
-    ///     var exampleGroup = new AzureAD.Group("exampleGroup", new()
+    ///     var example = new AzureAD.Group("example", new()
```
Are these intentional? Seems like all the resources here have the same name now.

t0yv0 replied:
Yeah we discussed it with platform team, it'd be nice to have it the other way but not blocking on it.

VenelinMartinov commented on provider/cmd/pulumi-resource-azuread/schema.json:
```diff
@@ -8735,181 +8735,180 @@
+                        "type": "string"
                     },
                     "webs": {
-                        "type": "array",
```
Do you know what happened to all of these? I don't see anything in your changes which would cause this.

t0yv0 replied:
These aren't semantic changes. IDK, something changed in Go struct devs for JSON serde? 

```
                     "users": {
-                        "type": "array",
+                        "description": "A list of users. Each `user` object provides the attributes documented below.\n",
                         "items": {
                             "$ref": "#/types/azuread:index/getUsersUser:getUsersUser"
                         },
-                        "description": "A list of users. Each `user` object provides the attributes documented below.\n"
+                        "type": "array"
                     }
```

VenelinMartinov commented on sdk/python/pulumi_azuread/user.py:
```diff
@@ -1493,10 +1493,10 @@ def __init__(__self__,
         import pulumi_azuread as azuread
 
         example = azuread.User("example",
+            user_principal_name="jdoe@example.com",
```
There's quite a few of these reorders - do you know what caused them? Will this happen only once or is the order no longer stable?

iwahbe replied:
The order is stable. It should only change once.
